bdk_chain/src/keychain/keychain_tracker.rs

```diff
@@ -121,26 +121,30 @@ where
     ///
     /// **Warning**: This function modifies the internal state of the tracker. You are responsible
     /// for persisting these changes to disk if you need to restore them.
-    pub fn insert_checkpoint(
-        &mut self,
+    pub fn insert_checkpoint_changeset(
+        &self,
         block_id: BlockId,
-    ) -> Result<bool, sparse_chain::InsertCheckpointErr> {
-        self.chain_graph.insert_checkpoint(block_id)
+    ) -> Result<KeychainChangeSet<K, P>, chain_graph::UpdateFailure<P>> {
+        Ok(KeychainChangeSet {
+            chain_graph: self.chain_graph.insert_checkpoint_changeset(block_id)?,
+            ..Default::default()
+        })
     }
 
     /// Inserts a transaction into the inner [`ChainGraph`] and optionally into the inner chain at
     /// `position`.
     ///
     /// **Warning**: This function modifies the internal state of the chain graph. You are
     /// responsible for persisting these changes to disk if you need to restore them.
-    pub fn insert_tx(
-        &mut self,
+    pub fn insert_tx_changeset(
+        &self,
         tx: Transaction,
-        position: Option<P>,
-    ) -> Result<bool, sparse_chain::InsertTxErr> {
-        let changed = self.chain_graph.insert_tx(tx.clone(), position)?;
-        self.txout_index.scan(&tx);
-        Ok(changed)
+        pos: P,
+    ) -> Result<KeychainChangeSet<K, P>, chain_graph::UpdateFailure<P>> {
```

Comment:
The error should not change. `InsertTxErr` correctly captured the things that could go wrong.